include/foxy/client_session.hpp

```diff
@@ -53,35 +53,26 @@ struct client_session : public session
 
   explicit client_session(boost::asio::io_context& io, session_opts opts = {});
 
-  // async_connect wraps asio::async_connect method and will invoke
-  // asio::async_connect with the specified host and service parameters
-  // async_connect is SSL-aware and will automatically perform the async
-  // handshake for the user, using the SSL context found in the session's opts
+  // async_connect wraps asio::async_connect method and will invoke asio::async_connect with the
+  // specified host and service parameters async_connect is SSL-aware and will automatically perform
+  // the async handshake for the user, using the SSL context found in the session's opts
   //
   template <class ConnectHandler>
-  auto async_connect(
-    std::string host,
-    std::string service,
-    ConnectHandler&&
-      handler) & -> BOOST_ASIO_INITFN_RESULT_TYPE(ConnectHandler,
-                                                  void(
-                                                    boost::system::error_code,
-                                                    boost::asio::ip::tcp::
-                                                      endpoint));
-
-  // async_request will serialize the provided Request to the connected remote
-  // and then use the provided ResponseParser to parse the response
-  // The user is expected to retrieve the underlying message via the Beast
-  // http::parser interface
+  auto
+  async_connect(std::string host, std::string service, ConnectHandler&& handler) & -> return_t<
+    ConnectHandler,
+    void(boost::system::error_code, boost::asio::ip::tcp::endpoint)>;
+
+  // async_request will serialize the provided Request to the connected remote and then use the
+  // provided ResponseParser to parse the response The user is expected to retrieve the underlying
+  // message via the Beast http::parser interface
   //
   template <class Request, class ResponseParser, class RequestHandler>
```

Comment:
Passing the entire Request and ResponseParser objects as individual unconstrained types is not a great idea, they should be parameterized on the class template, or else they need to be constrained (and Beast doesn't have the type traits for that).